File: vscode-pax-feedback/src/providers/facade.ts
/**
 * Provider facade for selecting and delegating to appropriate provider
 */

import * as vscode from 'vscode';
import { Provider, ProviderConfig } from './interface';
import { UniversalProvider } from './universal';

export class ProviderFacade {
  private provider: Provider | null = null;
  private config: ProviderConfig;

  constructor(config: ProviderConfig) {
    this.config = config;
  }

  /**
   * Auto-detect and activate the appropriate provider
   */
  async activate(context: vscode.ExtensionContext): Promise<void> {
    const configuredProvider = vscode.workspace
      .getConfiguration('pax.feedback')
      .get<string>('provider', 'auto');

    if (configuredProvider !== 'auto') {
      // User explicitly selected provider
      this.provider = await this.createProvider(configuredProvider, context);
    } else {
      // Auto-detection priority: Copilot > Cursor > Codex > Universal
      this.provider = await this.detectProvider(context);
    }

    if (this.provider) {
      await this.provider.activate();
      console.log(`PAX Feedback: Activated provider: ${this.provider.name}`);
    } else {
      throw new Error('No provider available');
    }
  }

  /**
   * Deactivate the current provider
   */
  async deactivate(): Promise<void> {
    if (this.provider) {
      await this.provider.deactivate();
      this.provider = null;
    }
  }

  /**
   * Get the active provider
   */
  getProvider(): Provider | null {
    return this.provider;
  }

  /**
   * Auto-detect available provider
   */
  private async detectProvider(
    context: vscode.ExtensionContext
  ): Promise<Provider> {
    // Check for Copilot extension
    const copilotProvider = await this.createProvider('copilot', context);
    if (await copilotProvider.isAvailable()) {
      return copilotProvider;
    }

    // Check for Cursor extension
    const cursorProvider = await this.createProvider('cursor', context);
    if (await cursorProvider.isAvailable()) {
      return cursorProvider;
    }

    // Check for Codex API
    const codexProvider = await this.createProvider('codex', context);
    if (await codexProvider.isAvailable()) {
      return codexProvider;
    }

    // Default to universal provider
    return this.createProvider('universal', context);
  }

  /**
   * Create provider instance by name
   */
  private async createProvider(
    name: string,
    context: vscode.ExtensionContext
  ): Promise<Provider> {
    switch (name) {
      case 'copilot':
        // TODO: Implement CopilotProvider (Phase 6)
        throw new Error('Copilot provider not yet implemented');

      case 'codex':
        // TODO: Implement CodexProvider (Phase 6)
        throw new Error('Codex provider not yet implemented');

      case 'cursor':
        // TODO: Implement CursorProvider (Phase 6)
        throw new Error('Cursor provider not yet implemented');

      case 'universal':
      default:
        return new UniversalProvider(this.config, context);
    }
  }
}

File: skills/tools/capturing-events/providers/facade.py
"""Provider facade for assistant-agnostic event capture."""

import os
import sys
from typing import Optional

try:
    from universal import UniversalProvider
except ImportError:
    from .universal import UniversalProvider


class ProviderDetector:
    """Detect which provider/assistant is available."""

    @staticmethod
    def detect() -> str:
        """
        Detect active provider.

        Returns:
            Provider name: "copilot", "codex", "cursor", or "universal" (default)
        """
        # Check for GitHub Copilot in VS Code environment
        if "GITHUB_COPILOT_AGENT" in os.environ:
            return "copilot"

        # Check for Codex (OpenAI API key)
        if "OPENAI_API_KEY" in os.environ:
            return "codex"

        # Check for Cursor editor environment
        if "CURSOR" in os.environ or "CURSOR_PATH" in os.environ:
            return "cursor"

        # Check for VS Code environment (Copilot fallback)
        if "TERM_PROGRAM" in os.environ and "vscode" in os.environ.get(
            "TERM_PROGRAM", ""
        ).lower():
            return "copilot"

        # Default to universal (workspace-only) provider
        return "universal"


class ProviderFacade:
    """Facade for provider-agnostic event capture."""

    def __init__(self, provider: Optional[str] = None):
        """
        Initialize provider facade.

        Args:
            provider: Explicit provider name or None to auto-detect.
                     Valid values: "universal", "copilot", "codex", "cursor"
        """
        if provider is None:
            provider = ProviderDetector.detect()

        self.provider_name = provider
        self._provider = self._create_provider(provider)

    def _create_provider(self, provider: str):
        """
        Create provider instance.

        Args:
            provider: Provider name

        Returns:
            Provider instance

        Raises:
            ValueError: If provider is unknown
        """
        if provider == "universal":
            return UniversalProvider()
        elif provider == "copilot":
            # TODO: Implement Copilot provider (wi-004)
            return UniversalProvider()  # Fall back to universal
        elif provider == "codex":
            # TODO: Implement Codex provider (wi-004)
            return UniversalProvider()  # Fall back to universal
        elif provider == "cursor":
            # TODO: Implement Cursor provider (wi-004)
            return UniversalProvider()  # Fall back to universal
        else:
            raise ValueError(
                f"Unknown provider: {provider}. "
                'Valid options: "universal", "copilot", "codex", "cursor"'
            )

    def capture_file_event(self, event_type: str, filepath: str) -> dict:
        """
        Capture file system event.

        Args:
            event_type: "file_create", "file_modify", or "file_delete"
            filepath: Path to file

        Returns:
            Event dict
        """
        return self._provider.capture_file_event(event_type, filepath)

    def capture_terminal_event(
        self, event_type: str, command: str, output: str = "", error: str = ""
    ) -> dict:
        """
        Capture terminal event.

        Args:
            event_type: "terminal_execute", "terminal_output", or "terminal_error"
            command: Command executed
            output: Command output (if any)
            error: Error output (if any)

        Returns:
            Event dict
        """
        return self._provider.capture_terminal_event(event_type, command, output, error)

    def capture_diagnostic_event(
        self, event_type: str, file: str, line: int, message: str
    ) -> dict:
        """
        Capture diagnostic event.

        Args:
            event_type: "diagnostic_error", "diagnostic_warning", or "diagnostic_info"
            file: File path with diagnostic
            line: Line number
            message: Diagnostic message

        Returns:
            Event dict
        """
        return self._provider.capture_diagnostic_event(event_type, file, line, message)

    def capture_skill_event(self, event_type: str, skill_name: str, status: str) -> dict:
        """
        Capture skill invocation event.

        Args:
            event_type: "skill_invoke", "skill_complete", or "skill_error"
            skill_name: Name of the skill
            status: Status message

        Returns:
            Event dict
        """
        return self._provider.capture_skill_event(event_type, skill_name, status)
